[DOC-2727] Custom cache paths, clarify storage (#893)
harness/developer-hub: docs/continuous-integration/use-ci/caching-ci-data/cache-intelligence.md

The cache retention window is 15 days, which resets whenever the cache is updated.

## Enable Cache Intelligence

To enable Cache Intelligence on a CI Build stage, add the following lines to the `stage: spec:` in your pipeline's YAML:

```yaml
caching:
  enabled: true
```

For example:

```yaml
    - stage:
        name: Build Jhttp
        identifier: Build_Jhttp
        type: CI
        spec:
          caching:
            enabled: true
          cloneCodebase: true
...
```

### Customize cache paths

Cache Intelligence stores the data to be cached in the `/harness` directory by default. You can use `paths` to specify a list of locations to be cached. This is useful if:

* Cache Intelligence is not supported for your build tool.
* You have customized cache locations, such as with `yarn config set cache-folder`.

Add the `paths` list to your pipeline's YAML, for example:

```yaml
    - stage:
        name: Build
        identifier: Build
        type: CI
        spec:
          caching:
            enabled: true
            paths:
              - /harness/node-modules

>               - /my_cache_directory/module_cache1

Reviewer: Remove /my_cache_directory

here because that's for the next example - outside /harness

Author: already merged so fixing on separate pr: https://github.com/harness/developer-hub/pull/897

There are several ways to configure caching, such as save/restore cache steps and mounting volumes, which require users to enable caching and manage the cache itself.

With Cache Intelligence, Harness can automatically cache and restore common dependencies, making it easy to start using caching in your pipelines. You don't need to bring your own storage, because we store the cache in Harness Cloud, our hosted environment.

## Supported build infrastructures and tools

Currently, Cache Intelligence is available only when using [Harness Cloud](/docs/continuous-integration/ci-quickstarts/hosted-builds-on-virtual-machines-quickstart), the Harness-hosted build environment.

Currently, Cache Intelligence supports Bazel, Maven, Gradle, Yarn, and Node build tools, if the dependencies are stored in the default location for the tool used.

If you are using a different build tool or a non-default cache location, you can still leverage our cache storage by [specifying the location(s) to cache](#customize-cache-paths).

## Cache storage

> Harness Cloud allows up to 2GB of cache storage per account. All pipelines in the account use the same cache storage, and each build tool has a unique cache key that is used to restore the appropriate cache data at runtime. This is similar to how you would specify the `cache-key` in **Save Cache** and **Restore Cache** steps.

Reviewer: I think the last sentence is not needed.

sidebar_position: 20
---

Caching dependencies is an effective way to speed up your build.

> Modern continuous integration systems execute pipelines inside ephemeral environments, such as a container or virtual machine, that is provisioned solely for pipeline execution and is not reused from a prior execution. As builds often require downloading and installing many dependencies, such as libraries and other software components, caching these dependencies can save a significant amount of time.

Reviewer 2: ```suggestion
Modern continuous integration systems execute pipelines inside ephemeral environments that are provisioned solely for pipeline execution and are not reused from prior executions. As builds often require downloading and installing many library and software dependencies, caching these dependencies can save a significant amount of time.
```

Reviewer 2: you had 2 "such as" in a row so I'm just suggesting that you remove them. Our audience should know what ephemeral environments and dependencies are.

I tend to do intros in a problem -> solution format. To me, the key points are:

Ephemeral environments provide a clean slate for each build and test run, ensuring that the environment is consistent and free of any artifacts or configurations from previous runs. However, this also means that each run must start from scratch, which can be time-consuming and resource-intensive.

Caching solves this problem by storing frequently used data or artifacts in a cache, which can be quickly retrieved instead of having to be regenerated or downloaded each time. By caching dependencies, packages, and artifacts, builds and tests can be executed more quickly, and the overall time to complete a CI pipeline can be reduced.

There are several ways to configure caching, such as save/restore cache steps and mounting volumes, which require users to enable caching and manage the cache itself.

> With Cache Intelligence, Harness can automatically cache and restore common dependencies, making it easy to start using caching in your pipelines. You don't need to bring your own storage, because we store the cache in Harness Cloud, our hosted environment.

Reviewer 2: ```suggestion
With Harness Cache Intelligence, you can automatically cache and restore common dependencies, making it easy to start using caching in your pipelines. You don't need to bring your own storage, because we store the cache in our hosted environment, Harness Cloud.
```

The cache retention window is 15 days, which resets whenever the cache is updated.

## Enable Cache Intelligence

To enable Cache Intelligence on a CI Build stage, add the following lines to the `stage: spec:` in your pipeline's YAML:

```yaml
caching:
  enabled: true
```

For example:

```yaml
    - stage:
        name: Build Jhttp
        identifier: Build_Jhttp
        type: CI
        spec:
          caching:
            enabled: true
          cloneCodebase: true
...
```

### Customize cache paths

Cache Intelligence stores the data to be cached in the `/harness` directory by default. You can use `paths` to specify a list of locations to be cached. This is useful if:

* Cache Intelligence is not supported for your build tool.
* You have customized cache locations, such as with `yarn config set cache-folder`.

Add the `paths` list to your pipeline's YAML, for example:

```yaml
    - stage:
        name: Build
        identifier: Build
        type: CI
        spec:
          caching:
            enabled: true
            paths:
              - /harness/node-modules
              - /my_cache_directory/module_cache1
          cloneCodebase: true
...
```

> If a path you want to cache is outside the `/harness` directory, you must also specify this as a shared path. In the Visual editor, you can add **Shared Paths** in the stage's **Overview** settings. In the YAML editor, add a list of `sharedPaths` to the `stage: spec:`, for example:

Reviewer 2: no screenshots? Just YAML?
You could do tabs for Pipeline Studio, YAML, API. You can see examples in my PR here:

https://hdh.pr.harness.io/pr-878/docs/continuous-delivery/cd-services/k8s-services/kubernetes-services

>         description: ""

Reviewer: Remove or add some string

        type: CI
        spec:
          caching:
            enabled: true
          cloneCodebase: true
...
```

### Customize cache paths

> Cache Intelligence stores cache data in the `/harness` directory by default. You can specify cache locations within this directory or add cache locations outside this directory. To do this, add a list of `paths` to your pipeline's YAML, for example:

Reviewer: stores the data to be cached. 

You can use `paths` to specify a list of locations to be cached. This is useful if:
- Cache Intelligence is not supported for your build tool
- You have customized cache locations (eg: `yarn config set cache-folder`)

If the directories to be cached are outside the `/harness` directory, they should also be added to `sharedPaths` to be able to access them across steps.

```yaml
    - stage:
        name: Build
        identifier: Build
        type: CI
        spec:
          caching:
            enabled: true
            paths:
              - /harness/node-modules
              - /my-cache-directory/module-cache-1
          cloneCodebase: true
...
```

If your cache path is outside the `/harness` directory, you must specify also specify this as a shared path. In the Visual editor, you can add **Shared Paths** in the stage's **Overview** settings. In the YAML editor, add a list of `sharedPaths` to the `stage: spec:`, for example:

```yaml
    - stage:
        name: Build
        identifier: Build
        type: CI
        spec:
          caching:
            enabled: true
            paths:

>               - /harness/node-modules

Reviewer: node_modules
